MM-51231 Replace react-custom-scrollbars with react-simplebar

hmhealey opened the pull request:
#### Summary
This was a relatively straightforward replacement except in a few places where I really had to fight with our CSS, most notably:
1. To use this for team sidebar, channel sidebar, and system console sidebar, I had to make sure those were sized using flexbox and remove `overflow: visible` which their parents had for some reason. I assume most of the stuff I changes for that dates back to pre-flexbox days.
2. To use that in the modals that use it, I had to fight a bunch with the CSS for `.more-modal__list` since we use that in far too many different ways (with/without `.multi-select-wrapper`, with/without `.filtered-user-list`, and inside a React Select popout) that all contradict each other and need different `overflow` values to scroll or not. I tried to simplify that a bit, but it'll require a fair bit more work to unravel that.

As for the new library, I also wrapped it in a new `Scrollbars` component which abstracts that away and can also replace the repeated logic that we had used for `react-custom-scrollbars`.

#### Ticket Link
MM-51231

#### Screenshots
-|Before|After
-|-|-
Channel Sidebar|<img width="1281" height="693" alt="2a channel LHS before" src="https://github.com/user-attachments/assets/6c4f66fb-636b-4784-acdd-fd384e9f5ce7" />|<img width="1280" height="693" alt="2b channel LHS after" src="https://github.com/user-attachments/assets/8a908fca-6b73-4879-ac8d-2ab30871794a" />
Team Sidebar|<img width="1280" height="693" alt="3a team LHS before" src="https://github.com/user-attachments/assets/9f84f824-21e1-4195-91bc-381620c162e7" />|<img width="1281" height="693" alt="3b team LHS after" src="https://github.com/user-attachments/assets/ea48d630-0556-4997-ad0e-870237ea57f6" />
Search|<img width="1281" height="693" alt="1a search before" src="https://github.com/user-attachments/assets/3d2e8fe2-8231-4418-bc73-991d887ae70e" />|<img width="1281" height="693" alt="1b search after" src="https://github.com/user-attachments/assets/1221477d-20b4-414d-aefc-133de3dfc031" />
System Console|<img width="1280" height="693" alt="4a system console before" src="https://github.com/user-attachments/assets/9b5b2211-e6d4-4964-bd0b-a9a11a545724" />|<img width="1280" height="693" alt="4b system console after" src="https://github.com/user-attachments/assets/0375bdbb-7061-4967-b1ad-3eade497693c" />

#### Release Note
```release-note
Updated library used for customizing scrollbars
```

hmhealey committed:
Add simplebar-react at latest version supported by React 17

hmhealey committed:
Remove react-custom-scrollbars

hmhealey committed:
Prevent sidebars and centre channel from overflowing the page vertically

hmhealey committed:
Add Scrollbars component and replace react-custom-scrollbars

hmhealey committed:
Fix usage of Scrollbars in modals

The various way we reuse .more-modal, .more-modal__list, and
.filtered-user-list in slightly different ways really made painful. We
should revisit that at some point when we get the chance since those
classes are used in a bunch of contradictory ways like using
.more-modal__list in a popover list in Multiselect versus as part of the
modal in the Browse Channel modal. It also doesn't help that some modals
use Scrollbars and others just use `overflow: auto` on either the whole
modal or on the .more-modal__list.

hmhealey committed:
Fix dragging the LHS scrollbar also dragging the channel behind it

hmhealey committed:
Fix scrolling in Browse Channels modal

hmhealey committed:
Fix results in user group modals not scrolling

hmhealey commented:
@abhijit-singh I added you for review since there's some minor changes to the position of the scrollbar in the places we customize it using this library (mostly the 4 places I screenshotted above) and because I moved some extra padding at the top of the LHS. Let me know if you want any further changes right now, but I might defer them since this blocks React 18.

---

@yasserfaraazkhan For reviewing this, in addition to the places screenshotted above, please take a quick look at any modals you can think of that might scroll on smaller screen sizes (usually ones that list users or other things) since I discovered how many different-looking modals share some of this CSS. I found and fixed a few issues in the Browse Channels, More DMs, and User Groups modals already, but if you have any other ideas of ones that might break, I'd appreciate the help.

hmhealey commented on e2e-tests/cypress/tests/integration/channels/channel_settings/more_unreads_position_with_scroll_spec.ts:
```diff
@@ -52,48 +52,41 @@ describe('Channel settings', () => {
+        cy.get('#unreadIndicatorBottom').click();
 
-        // # Post as another user in a channel near the bottom of the list, scroll channels list to view it (should be in bold)
-        cy.apiGetChannelByName(myTeam.name, channelNames[lastChannelIndex]).then(({channel}) => {
-            cy.postMessageAs({
-                sender: otherUser,
-                message: 'Bleep bloop I am a robot',
-                channelId: channel.id,
-            });
+        // * The list should have scrolled down to the bottom
+        cy.get(`#sidebarItem_${channelNames[firstChannelIndex]}`).should('not.be.visible');
+        cy.get(`#sidebarItem_${channelNames[lastChannelIndex]}`).should('be.visible');
 
-            // # Scroll down in channels list until last created channel is visible
-            cy.get(`#sidebarItem_${channelNames[firstChannelIndex]}`).scrollIntoView({duration: TIMEOUTS.TWO_SEC});
-            cy.get('.scrollbar--view').scrollTo('top');
-        });
```
I ended up rewriting this test since it seems like the original test was a bit flaky and then got weirder as people attempted to fix the flakiness. For example, line 73 clicks "More Unreads" to scroll to the top, and then line 84 scrolls the first channel into view, and then line 85 scrolls to the top, but the comment on line 83 says we're scrolling down to the last channel 🫤

hmhealey commented on webapp/channels/src/components/admin_console/access_control/modals/job_details/searchable_sync_job_channel_list.tsx:
```diff
@@ -239,7 +239,7 @@ const SearchableSyncJobChannelList = (props: Props) => {
 
     const input = (
-        <div className='filter-row filter-row--full'>
+        <div className='filter-row'>
```
This change wasn't strictly necessary for this, but it seems like these `filter-row` and `filter-row--full` ended up behaving exactly the same, so I got rid of the redundant one

hmhealey commented on webapp/channels/src/components/common/scrollbars.scss:
```diff
@@ -0,0 +1,21 @@
+
+[data-simplebar] {
+    // Ensures that scrollbar divs never overflow their flexbox parents
+    min-height: 0;
```
This is some weird CSS thing I learned about. If you have a flexbox with `overflow: hidden` or `overflow: auto` and an overflowing child in it, the child defaults to `min-height: auto` which causes the parent flexbox to expand to fit that child. If you set it to `min-height: 0`, the child will actually be cut off by its parent so it can scroll

hmhealey commented on webapp/channels/src/sass/layout/_sidebar-left.scss:
```diff
@@ -4,25 +4,6 @@
 $sidebarHeightTransitionDuration: 0.18s;
 $sidebarOpacityAnimationDuration: 0.15s;
 
-#TeamSidebar {
```
As best I can tell, this hasn't actually been used for a while

hmhealey committed:
Update snapshots and fix tests

hmhealey commented:
I'll look into why those E2E tests are failing when I'm back in September. Feel free to still review since I expect the changes required for those to be minor

larkox commented on webapp/channels/src/components/post_edit_history/post_edit_history.scss:
```diff
@@ -1,4 +1,4 @@
-.sidebar-right__edit-post-history .scrollbar--view {
+.sidebar-right__edit-post-history .simplebar-content {
```
I love that we are putting the new scrollbars in a new component with minimum interface. But it seems we are bleeding some library specifics into this file. Can we avoid it?

(I say it because `simplebar` sounds like the new library :P )

hmhealey replied:
I tried to do that wherever I could, but our CSS is so awkward in some of these areas that I couldn't always do that. I would've liked to remove that entirely similar to the `[data-simplebar]` styles I added to the new component, but there was some reason that it didn't work here

larkox commented on webapp/channels/src/components/search_results/search_results.tsx:
```diff
@@ -411,13 +383,7 @@ const SearchResults: React.FC<Props> = (props: Props): JSX.Element => {
-                renderThumbHorizontal={renderThumbHorizontal}
-                renderThumbVertical={renderThumbVertical}
-                renderView={renderView}
+                color='--center-channel-color-rgb'
```
It seems off that we have to be specific about the color here. But 0/5.

larkox commented on webapp/channels/src/components/common/scrollbars.tsx:
```diff
@@ -0,0 +1,51 @@
+        }
+    }, [onScroll]);
+
+    const mergedRef = useMergeRefs<HTMLElement>([ref, setScrollRef]);
```
Probably overkill, but we could use here the `useImperativeHandle`. That way, we can reduce and control the interface of this. That way, in many places where you need to do changes like `scrollToTop() -> scrollTo({top: 0})` we could just not make those changes, by exposing those things.

This way, if we ever change this library to something else, we just need to change the body of the imperative handler, and not everywhere in the code.

hmhealey replied:
I like that idea where if we were using those refs in more places, and if we needed to do more complicated things with them, but I generally prefer using the vanilla JS methods instead of using an imperative handle in cases like this

larkox commented on webapp/channels/src/sass/components/_modal.scss:
```diff
@@ -762,7 +758,7 @@
     }
 
-    >div {
+    >div:not([data-simplebar]) {
```
More implementation details of the new scroll bars bleeding into other parts of the code.

larkox reviewed (approved):
Looks great! There are a couple of comments, but they are not blocking. I leave whether to address them or not to your discretion.

abhijit-singh reviewed (approved):
Looks good to me, thanks @hmhealey!

hmhealey committed:
Merge branch 'master' into MM-51231_scrollbars

hmhealey committed:
Prevent focusing simplebar divs (accessibility_sidebar_spec.ts)

hmhealey committed:
Fix admin LHS still being able to scroll (openid_spec.ts)

hmhealey committed:
Migrate Team Members modal to GenericModal and the new menu (manage_members_spec.js)

The menu was being cut off yet again by the contents of the modal, so I
migrated that to the new menu to portal it out of that. Unfortunately,
that caused the infinite recursion bug between MUI and React Bootstrap's
focus trap logic, so I had to also migrate the modal to GenericModal to
fix that.

hmhealey committed:
Update snapshot and fix test

hmhealey committed:
Update snapshot and fix test

hmhealey committed:
Fix another E2E test

hmhealey committed:
More snapshots

hmhealey committed:
Merge remote-tracking branch 'origin/master' into MM-51231_scrollbars

tnir commented on webapp/channels/package.json:
- `@types/react-custom-scrollbars` npm package dep should be removed in this PR, but followed up in #36818.
